File: Examples/AirGapDeposition/AirGapDeposition.cpp
#include <iostream>

#include <lsAdvect.hpp>
#include <lsBooleanOperation.hpp>
#include <lsDomain.hpp>
#include <lsExpand.hpp>
#include <lsMakeGeometry.hpp>
#include <lsPrune.hpp>
#include <lsToMesh.hpp>
#include <lsToSurfaceMesh.hpp>
#include <lsVTKWriter.hpp>

/**
  2D Example showing how to use the library for topography
  simulation, by creating a trench geometry. A layer of a different material is
  then grown directionally on top. \example AirGapDeposition.cpp
*/

using NumericType = float;

// implement own velocity field
class velocityField : public lsVelocityField<NumericType> {
public:
  NumericType
  getScalarVelocity(const std::array<NumericType, 3> & /*coordinate*/,
                    int /*material*/,
                    const std::array<NumericType, 3> &normalVector,
                    unsigned long /*pointId*/) {
    // velocity is proportional to the normal vector
    NumericType velocity =
        std::abs(normalVector[0]) + std::abs(normalVector[1]);
    return velocity;
  }

  std::array<NumericType, 3>
  getVectorVelocity(const std::array<NumericType, 3> & /*coordinate*/,
                    int /*material*/,
                    const std::array<NumericType, 3> & /*normalVector*/,
                    unsigned long /*pointId*/) {
    return std::array<NumericType, 3>({});
  }
};

int main() {

  constexpr int D = 2;
  omp_set_num_threads(2);

  NumericType extent = 30;
  NumericType gridDelta = 0.5;

  hrleCoordType bounds[2 * D] = {-extent, extent, -extent, extent};
  lsDomain<NumericType, D>::BoundaryType boundaryCons[D];
  boundaryCons[0] = lsDomain<NumericType, D>::BoundaryType::REFLECTIVE_BOUNDARY;
  boundaryCons[1] = lsDomain<NumericType, D>::BoundaryType::INFINITE_BOUNDARY;

  auto substrate = lsSmartPointer<lsDomain<NumericType, D>>::New(
      bounds, boundaryCons, gridDelta);

  NumericType origin[2] = {0., 0.};
  NumericType planeNormal[2] = {0., 1.};

  {
    auto plane =
        lsSmartPointer<lsPlane<NumericType, D>>::New(origin, planeNormal);
    lsMakeGeometry<NumericType, D>(substrate, plane).apply();
  }

  {
    std::cout << "Extracting..." << std::endl;
    auto mesh = lsSmartPointer<lsMesh<NumericType>>::New();
    lsToSurfaceMesh<NumericType, D>(substrate, mesh).apply();
    lsVTKWriter<NumericType>(mesh, "plane.vtk").apply();
  }

  {
    // create layer used for booling
    std::cout << "Creating box..." << std::endl;
    auto trench = lsSmartPointer<lsDomain<NumericType, D>>::New(
        bounds, boundaryCons, gridDelta);
    NumericType xlimit = extent / 6.;
    NumericType minCorner[D] = {-xlimit, -25.};
    NumericType maxCorner[D] = {xlimit, 1.};
    auto box = lsSmartPointer<lsBox<NumericType, D>>::New(minCorner, maxCorner);
    lsMakeGeometry<NumericType, D>(trench, box).apply();

    {
      std::cout << "Extracting..." << std::endl;
      auto mesh = lsSmartPointer<lsMesh<NumericType>>::New();
      lsToMesh<NumericType, D>(trench, mesh).apply();
      lsVTKWriter<NumericType>(mesh, "box.vtk").apply();
    }

    // Create trench geometry
    std::cout << "Booling trench..." << std::endl;
    lsBooleanOperation<NumericType, D>(
        substrate, trench, lsBooleanOperationEnum::RELATIVE_COMPLEMENT)
        .apply();
  }

  // Now grow new material

  // create new levelset for new material, which will be grown
  // since it has to wrap around the substrate, just copy it
  std::cout << "Creating new layer..." << std::endl;
  auto newLayer = lsSmartPointer<lsDomain<NumericType, D>>::New(substrate);

  auto velocities = lsSmartPointer<velocityField>::New();

  std::cout << "Advecting" << std::endl;
  lsAdvect<NumericType, D> advectionKernel;

  // the level set to be advected has to be inserted last
  // the other could be taken as a mask layer for advection
  advectionKernel.insertNextLevelSet(substrate);
  advectionKernel.insertNextLevelSet(newLayer);

  advectionKernel.setVelocityField(velocities);
  advectionKernel.setIgnoreVoids(true);

  // Now advect the level set 50 times, outputting every
  // advection step. Save the physical time that
  // passed during the advection.
  NumericType passedTime = 0.;
  unsigned numberOfSteps = 60;
  for (unsigned i = 0; i < numberOfSteps; ++i) {
    advectionKernel.apply();
    passedTime += advectionKernel.getAdvectedTime();

    std::cout << "\rAdvection step " + std::to_string(i) + " / "
              << numberOfSteps << std::flush;
    auto mesh = lsSmartPointer<lsMesh<NumericType>>::New();
    lsToSurfaceMesh<NumericType, D>(newLayer, mesh).apply();
    lsVTKWriter<NumericType>(mesh, "trench" + std::to_string(i) + ".vtk")
        .apply();
  }
  std::cout << std::endl;
  std::cout << "Time passed during advection: " << passedTime << std::endl;

  return 0;
}

File: Examples/AirGapDeposition/AirGapDeposition.py
import viennaLS2d as vls

## @example AirGapDeposition.py
#  Example showing how to use the library for topography
#  simulation, by creating a trench geometry. A layer of a different material is
#  then grown directionally on top.

class velocityField(vls.lsVelocityField):
  # coord and normalVec are lists with 3 elements
  # in 2D coord[2] and normalVec[2] are zero
  # getScalarVelocity must return a scalar
  def getScalarVelocity(self, coord, material, normal,  pointId):
    return abs(normal[0]) + abs(normal[1])

  def getVectorVelocity(self, coord, material, normal,  pointId):
    return (0,0,0)

extent = 30
gridDelta = 0.5

bounds = (-extent, extent, -extent, extent)
boundaryCons = (0, 1, 0) # 0 = reflective, 1 = infinite, 2 = periodic

# create level set
substrate = vls.lsDomain(bounds, boundaryCons, gridDelta)

# create plane
origin = (0,0,0)
planeNormal = (0,1,0)

vls.lsMakeGeometry(substrate, vls.lsPlane(origin, planeNormal)).apply()

print("Extracting")
mesh = vls.lsMesh()
vls.lsToSurfaceMesh(substrate, mesh).apply()
vls.lsVTKWriter(mesh, "plane.vtk").apply()

# create layer used for booling
print("Creating box...")
trench = vls.lsDomain(bounds, boundaryCons, gridDelta)
minCorner = (-extent / 6., -25.)
maxCorner = (extent / 6., 1.)
vls.lsMakeGeometry(trench, vls.lsBox(minCorner, maxCorner)).apply()

print("Extracting")
vls.lsToMesh(trench, mesh).apply()
vls.lsVTKWriter(mesh, "box.vtk").apply()

# Create trench geometry
print("Booling trench")
vls.lsBooleanOperation(substrate, trench, vls.lsBooleanOperationEnum.RELATIVE_COMPLEMENT).apply()

# Now grow new material

# create new levelset for new material, which will be grown
# since it has to wrap around the substrate, just copy it
print("Creating new layer...")
newLayer = vls.lsDomain(substrate)

velocities = velocityField()

print("Advecting")
advectionKernel = vls.lsAdvect()

# the level set to be advected has to be inserted last
# the other could be taken as a mask layer for advection
advectionKernel.insertNextLevelSet(substrate)
advectionKernel.insertNextLevelSet(newLayer)

advectionKernel.setVelocityField(velocities)
advectionKernel.setIgnoreVoids(True)

# Now advect the level set 50 times, outputting every
# advection step. Save the physical time that
# passed during the advection.
passedTime = 0
numberOfSteps = 60
for i in range(numberOfSteps):
  advectionKernel.apply()
  passedTime += advectionKernel.getAdvectedTime()

  print("Advection step {} / {}".format(i, numberOfSteps))

  vls.lsToSurfaceMesh(newLayer, mesh).apply()
  vls.lsVTKWriter(mesh, "trench{}.vtk".format(i)).apply()

print("Time passed during advection: {}".format(passedTime))
